src/libraries/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncHelpers.cs

```diff
@@ -0,0 +1,175 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Runtime.CompilerServices
+{
+    [System.ComponentModel.EditorBrowsable(System.ComponentModel.EditorBrowsableState.Never)]
+    [System.Diagnostics.CodeAnalysis.ExperimentalAttribute("SYSLIB5007", UrlFormat = "https://aka.ms/dotnet-warnings/{0}")]
+    public static partial class AsyncHelpers
+    {
+#if CORECLR
+        [BypassReadyToRun]
+        [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.Async)]
+        public static void AwaitAwaiter<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion
+        {
+            ref AsyncHelpers.RuntimeAsyncAwaitState state = ref AsyncHelpers.t_runtimeAsyncAwaitState;
+            Continuation? sentinelContinuation = state.SentinelContinuation;
+            if (sentinelContinuation == null)
+                state.SentinelContinuation = sentinelContinuation = new Continuation();
+
+            state.Notifier = awaiter;
+            AsyncHelpers.AsyncSuspend(sentinelContinuation);
+        }
+
+        [BypassReadyToRun]
+        [MethodImpl(MethodImplOptions.NoInlining | MethodImplOptions.Async)]
+        public static void UnsafeAwaitAwaiter<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
+        {
+            ref AsyncHelpers.RuntimeAsyncAwaitState state = ref AsyncHelpers.t_runtimeAsyncAwaitState;
+            Continuation? sentinelContinuation = state.SentinelContinuation;
+            if (sentinelContinuation == null)
+                state.SentinelContinuation = sentinelContinuation = new Continuation();
+
+            state.Notifier = awaiter;
+            AsyncHelpers.AsyncSuspend(sentinelContinuation);
+        }
+
+        // Marked intrinsic since JIT recognises the helper by name when doing optimizations.
+        [Intrinsic]
+        [BypassReadyToRun]
```

Comment:
Comment on `BypassReadyToRun` would be more helpful. Is it a permanent thing or just a temporary measure since R2R support is not yet implemented.

Reply:
I think it is because R2R support is not yet implemented, but @jakobbotsch would know better.

If all Async methods are excluded from R2R anyways, maybe this is not necessary?

Reply:
The `[NoInlining]` on some methods is because the method suspends explicitly with a manually created continuation. 
No locals or other state would be automatically added to the continuation, thus it is better if the containing method is never inlined.  

Using `AsyncSuspend` would generally require `NoInlining` for the containing method.
It seems it is worth a comment as well. I'll add a few words.

Reply:
`BypassReadyToRun` can likely be removed. Back when we had both prototypes and switched between them via environment variable this function had two different implementations depending on which one was selected, and this was just the simplest way to ensure we did not prejit this function into one of them.

Reply:
I've removed `BypassReadyToRun` and tests are passing. 

We exclude Async variants from tiering for now, that would also make then not R2R able. But these are not variants, so I guess R2R just works. No reason not to, I think. I did not verify for 100%, but seeing no failures is enough to remove `BypassReadyToRun`

Reply:
Added comments about `NoInline`